src/persist-client/src/batch.rs

```diff
@@ -187,17 +187,26 @@ where
     size_hint: usize,
     lower: Antichain<T>,
     max_ts: T,
+    blob_target_size: usize,
 
     shard_id: ShardId,
-    records: ColumnarRecordsVecBuilder,
     blob: Arc<dyn Blob + Send + Sync>,
     metrics: Arc<Metrics>,
 
+    current_part: Vec<((Vec<u8>, Vec<u8>), T, D)>,
```

Comment:
I mentioned this in the slack thread, but all these little allocations are bad for perf. if we're careful about lifetimes, should be possible to make all of those into slices. so have one big `Vec<u8>` that is cleared between parts and then then we collect things as `Vec<(&[u8], &[u8], T, D)>` or `BTreeMap<(&[u8], &[u8], T), D>`